cocos/animation/marionette/animation-graph-animation-clip-binding.ts

```diff
@@ -0,0 +1,349 @@
+}
+
+/**
+ * The context in which animation clips can evaluate during animation graph evaluation.
+ * Currently, the context is just the output pose.
+ */
+export type AnimationClipGraphEvaluationContext = Pose;
+
+/**
+ * A pose binding describes how to get/set part of a bound transform in animation graph.
+ * The `T` can be `Vec3`, `Quat` for now.
+ */
+export interface PoseBinding<T> {
+    /**
+     * Destroys this binding.
+     */
+    destroy(): void;
+
+    /**
+     * Sets the part's value.
+     * @param value The value.
+     * @param pose The pose.
+     */
+    setValue(value: T, pose: Pose): void;
+
+    /**
+     * Reads the part's value.
+     * @param pose The pose.
+     */
+    getValue(pose: Pose): Readonly<T>;
+}
+
+const CACHE_VEC3_GET_VALUE = new Vec3();
+
+const CACHE_QUAT_GET_VALUE = new Quat();
+
+/**
+ * The pose binding base is base class of all pose binding classes.
+ * It holds a transform handle.
+ */
+class PoseBindingBase {
+    constructor (transformHandle: TransformHandle) {
+        this._transformHandle = transformHandle;
+    }
+
+    /**
+     * Releases the held transform handle.
+     */
+    public destroy () {
+        this._transformHandle.destroy();
+    }
+
+    /**
+     * The held transform handle.
+     */
+    protected declare readonly _transformHandle: TransformHandle;
+}
+
+/**
+ * The pose position binding describes how to get/set the position of a bound transform in animation graph.
+ */
+class PosePositionBinding extends PoseBindingBase implements PoseBinding<Vec3> {
+    public setValue (value: Vec3, pose: Pose): void {
+        pose.transforms.setPosition(this._transformHandle.index, value);
+    }
+
+    public getValue (pose: Pose) {
+        return pose.transforms.getPosition(this._transformHandle.index, CACHE_VEC3_GET_VALUE) as Readonly<Vec3>;
+    }
+}
+
+/**
+ * The pose rotation binding describes how to get/set the rotation(in quaternion) of a bound transform in animation graph.
+ */
+class PoseRotationBinding extends PoseBindingBase implements PoseBinding<Quat> {
+    public setValue (value: Quat, pose: Pose): void {
+        pose.transforms.setRotation(this._transformHandle.index, value);
+    }
+
+    public getValue (pose: Pose) {
+        return pose.transforms.getRotation(this._transformHandle.index, CACHE_QUAT_GET_VALUE) as Readonly<Quat>;
+    }
+}
+
+/**
+ * The pose euler angles binding describes how to get/set the rotation(in euler angles) of a bound transform in animation graph.
+ */
+class PoseEulerAnglesBinding extends PoseBindingBase implements PoseBinding<Vec3> {
+    public setValue (value: Vec3, pose: Pose): void {
+        const quat = Quat.fromEuler(PoseEulerAnglesBinding._EULER_TO_QUAT_CACHE, value.x, value.y, value.z);
+        pose.transforms.setRotation(this._transformHandle.index, quat);
+    }
+
+    public getValue (pose: Pose) {
+        const q = pose.transforms.getRotation(this._transformHandle.index, CACHE_QUAT_GET_VALUE) as Readonly<Quat>;
```

Comment:
This may cause some bugs! Because we are storing quaternions, so this will cause us to get different values for setValue and getValue when we set them using eulerAngles. In particular, when the incoming eulerAngles are Vec3(90, 90, 90). Need to be careful here.

Reply:
Seems like there's no way out. It's a pitfall of `Node` itself.

`getValue` is invoked in "component-wise animation" context. For example user has only `.x` animation in the node's rotation.